src/compiler/moduleNameResolver.ts

```diff
@@ -1628,4 +1628,60 @@ namespace ts {
             return toSearchResult(resolved);
         }
     }
+
+    // Copied from `loadModuleFromSpecificNodeModulesDirectory`
+    function loadModuleFromPnpResolution(extensions: Extensions, packageDirectory: string, rest: string, state: ModuleResolutionState): Resolved | undefined {

```

Comment:
Do you see a way we could avoid copying all this code? I'm worried it will become fairly quickly out of sync with the original implementation 🤔

Reply:
The first thing I tried was delegating to `loadModuleFromSpecificNodeModulesDirectory`, but I wasn't sure `portal:` would work.

To use `loadModuleFromSpecificNodeModulesDirectory`, we need to resolve the package to a `..../node_modules/@scope/pkgname`, then parse that into `..../node_modules` and `@scope/pkgname`.  We can pass both of those strings to `loadModuleFromSpecificNodeModulesDirectory`.

Will it be compatible with `portal:` resolutions?  If a portal dep resolves to `/d/Personal-dev-projects/@foo`, where the workspace is named `@myscope/foo`, can we safely delegate this to `loadModuleFromSpecificNodeModulesDirectory`?  (this is an intentionally problematic example)

Reply:
Could we instead modify `loadModuleFromSpecificNodeModulesDirectory`? Something like this:

```ts
function loadModuleFromSpecificNodeModulesDirectory(extensions: Extensions, moduleName: string, nodeModulesDirectory: string, nodeModulesDirectoryExists: boolean, state: ModuleResolutionState): Resolved | undefined {
    const candidate = normalizePath(combinePaths(nodeModulesDirectory, moduleName));
    return loadModuleFromSpecificNodeModulesDirectory(extensions, moduleName, nodeModulesDirectory, nodeModulesDirectoryExists, state);
}


function loadModuleFromSpecificNodeModulesDirectoryImpl(extensions: Extensions, moduleName: string, nodeModulesDirectory: string, candidate: string, nodeModulesDirectoryExists: boolean, state: ModuleResolutionState): Resolved | undefined {
    // First look for a nested package.json, as in `node_modules/foo/bar/package.json`.
    let packageInfo = getPackageJsonInfo(candidate, !nodeModulesDirectoryExists, state);

    //...
}
```

Then we'd just have to use `loadModuleFromSpecificNodeModulesDirectoryImpl` instead of the problematic `loadModuleFromSpecificNodeModulesDirectory`.

Reply:
Yeah that should work.  I can make the change this afternoon.